src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -904,6 +903,105 @@ internal RunVersionOne CreateRun(Run v2Run)
             return run;
         }
 
+        // In SARIF v1, run.files was a dictionary, whereas in v2 it is an array.
+        // During the v2-to-v1 conversion, it is sometimes necessary to look up
+        // information from a v2 FileData object. Rather than having to search the
+        // v2 run.files array to find the required FileData object, we construct a
+        // dictionary from the array so we can access the required FileData object
+        // directly. It turns out that we need mappings in both directions: from
+        // array index to dictionary key, and vice versa.
+        private static void CreateFileKeyIndexMappings(
+            IList<FileData> v2Files,
+            out IDictionary<string, int> fileKeyToIndexDictionary,
+            out IDictionary<int, string> fileIndexToKeyDictionary)
+        {
+            fileKeyToIndexDictionary = null;
+            fileIndexToKeyDictionary = null;
+
+            if (v2Files != null)
+            {
+                fileKeyToIndexDictionary = new Dictionary<string, int>();
+                fileIndexToKeyDictionary = new Dictionary<int, string>();
+
+                for (int index = 0; index < v2Files.Count; ++index)
+                {
+                    FileData v2File = v2Files[index];
+                    string key = CreateFileDictionaryKey(v2File, v2Files);
+
+                    fileKeyToIndexDictionary[key] = index;
+                    fileIndexToKeyDictionary[index] = key;
+                }
+            }
+        }
+
+        // Given a v2 FileData object, synthesize a key that will be used to locate that object
+        // in the v2 file data dictionary. We will use the same key to locate the corresponding
+        // v1 FileData object when we create the v1 run.files dictionary.
+        //
+        // This method is necessary because, although ideally we would use the FileData
+        // object's URI as the dictionary key, it is possible for two FileData objects to
+        // have the same URI. For example:
+        //
+        //    1. Nested files with the same name might exist in two different containers.
+        //    2. There might be two files with the same URI but difference uriBaseIds.
+        //
+        // To deal with Case #1, we synthesize a key as follows:
+        //
+        //    <root file URI>#<nested file path>/<level 2 nested file path>/...
+        //
+        // To deal with Case #2, we would follow the chain or uriBaseIds, prepending
+        // the value of each one to the URI.
+        //
+        // WE DO NOT YET HANDLE CASE #2.
```

Comment:
This is the missing code and test case.

Reply:
I will handle this in a separate PR, after lunch, after I review your PR.

---
In reply to: [247375067](https://github.com/Microsoft/sarif-sdk/pull/1211#discussion_r247375067) [](ancestors = 247375067)